Add multicollection support

alarbada opened the pull request:
### Description

Adds support for multicollection writes.

raulb commented on destination/destination_integration_test.go:
```diff
@@ -41,11 +42,68 @@ var cfg = map[string]string{
+	stream3 := setupDestinationTest(ctx, con.client, is)
+
+	// wait for first stream to be created
+	err = con.Open(ctx)
```
I wonder if we should defer `con.Teardown(ctx)`

raulb reviewed (commented):
@alarbada Something I just noticed by looking at both the implementation and tests is the fact that we're using different configuration options for something that possibly be merged into one. I'm referring to when using multiple streams. 

In other connectors we took the approach to use the same key to determine whether it could use multiple collections via a go template, with a default value of the metadata field `opencdc.collection`. In other words, I'm wondering if you think would be possible to merge `streamName` (or is it `streamARN`?), `useMultiStreamMode` and `streamARNTemplate`) into just one configuration option so it's consistent with other connectors. 

What do you think?

alarbada commented:
@raulb I picked up the `streamName` usage from the old code. I wonder if I can just use `streamName` and leave `streamARN` behind..., let me check.

Also, last merge joins connection logic into `common/client.go`, that's the main thing.

alarbada commented:
And about joining `useMultiStreamMode`, `streamARNTemplate`and `streamARN`. If joining them together I guess that the configuration paths would end up like this:

- `streamARN` as a non go template string -> no problem, use single collection mode.
- `streamARN` empty -> streamARN is picked up from `opencdc.collection`. But what happens if none found? should we error?
- `streamARN` as a go template -> no problem, parse the streamARN from template execution. I guess here it is more valid to error out if no valid `streamARN` found.

So I separated them into three so that there was always a fallback. That's why I made `streamARN` required. But I don't know if that's a more comfortable path for the user. 

There's also the issue that if I use a single `streamARN` parameter, I don't know how to encode the default `{{ index .Metadata 'opencdc.collection' }}` behaviour. Having the three parameters allow me to express that in the configuration.

alarbada commented:
@raulb well I see in the kafka connector that we don't check whether the topic is or isn't in the record ([here's the reference, not checking for empty topic](https://github.com/ConduitIO/conduit-connector-kafka/blob/9bfe3e1f0b242d439a5157c1b1bc421fdc104201/destination/config.go#L148-L152)).

I can do the same, I see that it is not that much of an issue not finding the collection. I can maybe give a better error than letting kinesis error out on empty stream given.

alarbada commented:
Oh I completely forgot, I have to set the collection field in the source connector! That aswell for the pinecone connector.

alarbada commented:
Alright @raulb it is ready to be reviewed again.

Basically:
- removed much boilerplate from the integration test code, which was making it difficult to understand the tests.
- removed source integration test, as it was already covered by the acceptance tests.
- replaced `streamARN` by `streamName`. I believe the later is much more easier to use for the connector user.
- updated docs accordingly.
- simplify destination `.Configure`.

What do you think?

raulb commented on common/client.go:
```diff
@@ -0,0 +1,56 @@
+	sdk "github.com/conduitio/conduit-connector-sdk"
+)
+
+func NewClient(ctx context.Context, httpClient *http.Client, cfg Config) (*kinesis.Client, error) {
```
I think we should add a comment for this exported method. Even it could look self explanatory, it'd be nice to have for the generated go documentation.

alarbada replied:
You are right, and I should actually document why I'm passing a custom httpClient, doesn't make sense without looking at the internals.

raulb commented on source/config.go:
```diff
@@ -20,6 +20,9 @@ import "github.com/conduitio-labs/conduit-connector-kinesis/common"
 	common.Config
 
+	// StreamName is the name of the Kinesis Data Stream.
+	StreamName string `json:"streamName"`
```
```suggestion
	StreamName string `json:"streamName" validate:"required"`
```

raulb commented on destination/config.go:
```diff
@@ -21,6 +21,12 @@ type Config struct {
+	// [Go template](https://pkg.go.dev/text/template) that will be executed
+	// for each record to determine the stream name. By default, the stream
+	// name is the value of the `opencdc.collection` metadata field.
+	StreamName string `json:"streamName"`
```
```suggestion
	StreamName string `json:"streamName" default:"{{ index .Metadata \"opencdc.collection\" }}"`
```

alarbada replied:
I'm not sure about this one. If the standard is to get the collection from the `opencdc.collection` record field, why would we put the overhead of a go template execution? Specially since the `.Write()` method is in the hot path.
If we change zerolog calls from `.Info()` to `.Trace()`, I believe we should also not use a go template when getting the collection from `opencdc.collection`.

If we however want to still preserve the readability of having a `default:"{{ index .Metadata \"opencdc.collection\"  }}` in a struct tag, we could also just hardcode that template, like this:

```go
func parseStreamName(config Config, rec sdk.Record) (string, error) {
	if config.StreamName == "{{ index .Metadata \"opencdc.collection\" }}" {
		return rec.Metadata.GetCollection()
	}

	// parse and execute template
}
```

raulb commented on destination/batches.go:
```diff
@@ -0,0 +1,120 @@
+	ParseStreamName(sdk.Record) (string, error)
+}
+
+// fromColFieldParser parses the streamName from the opencdc.collection record
+// metadata field.
```
I think this comment is more appropriate with [`ParseStreamName`](https://github.com/conduitio-labs/conduit-connector-kinesis/pull/51/files#diff-451bf27366f9e28315c11b17e205da29d4bd6dcb7302f668e93b87efc71e3346R33)

raulb commented on destination/batches.go:
```diff
@@ -0,0 +1,120 @@
+func (p *fromColFieldParser) ParseStreamName(r sdk.Record) (string, error) {
+	streamName, err := r.Metadata.GetCollection()
+	if err != nil {
+		return "", fmt.Errorf("streamName not found in record %s", string(r.Key.Bytes()))
```
```suggestion
		return "", fmt.Errorf("streamName not found in record's collection metadata %s", string(r.Key.Bytes()))
```

raulb reviewed (commented):
@alarbada Looks pretty good! Left some things that should be simple to fix. Let me know!

alarbada commented on test/utils.go:
```diff
@@ -0,0 +1,133 @@
+	is.NoErr(err)
+
+	for i := 0; i < 10; i++ {
+		time.Sleep(time.Duration(100*math.Pow(2, float64(i))) * time.Millisecond)
```
The kinesis golang sdk creates streams asynchronously, and I haven't found any way to subscribe to stream status changes, so doing an exponential backoff is the simplest way I could come up with to wait for the stream to be ready.

raulb replied:
@alarbada I'm curious why waiting exponentially? i.e.: if we had a specific sleeping time would be sufficient for the stream to be ready>

alarbada replied:
@raulb well it's a common technique in network programming. The assumption is that if the service isn't ready in 1 second, it might be more probable that it is ready in 2 secs than in another sec.

I'm referring to [this](https://en.wikipedia.org/wiki/Exponential_backoff).


I saw it in multiple services to retry failed requests (eg. [openai](https://platform.openai.com/docs/guides/rate-limits), [google](https://cloud.google.com/memorystore/docs/redis/exponential-backoff), etc).

I also used exponential backoff in the pinecone connector. Their vector writes are async, so to assert a write in tests without making too many calls to the pinecone api this technique was useful.

The only drawback that I can think off is that it might make the tests slower, so that's why In the code example you referenced I used 100ms as a starting duration point (100, 200, 400, ... ms).

raulb replied:
@alarbada got it! thank you for the explanation.
